File: dev/Repeater/APITests/Common/OrientationBasedMeasures.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

using Windows.Foundation;
using Windows.UI.Xaml.Controls;

namespace Windows.UI.Xaml.Tests.MUXControls.ApiTests.RepeaterTests.Common
{
    public class OrientationBasedMeasures
    {
        public ScrollOrientation ScrollOrientation { get; set; }

        public bool IsVerical
        {
            get { return ScrollOrientation == ScrollOrientation.Vertical; }
        }

        public OrientationBasedMeasures(ScrollOrientation o)
        {
            ScrollOrientation = o;
        }

        public double Major(Size size)
        {
            return IsVerical ? size.Height : size.Width;
        }

        public double Minor(Size size)
        {
            return IsVerical ? size.Width : size.Height;
        }

        public double MajorSize(Rect rect)
        {
            return IsVerical ? rect.Height : rect.Width;
        }

        public double MinorSize(Rect rect)
        {
            return IsVerical ? rect.Width : rect.Height;
        }

        public double MajorStart(Rect rect)
        {
            return IsVerical ? rect.Top : rect.Left;
        }

        public double MajorEnd(Rect rect)
        {
            return IsVerical ? rect.Bottom : rect.Right;
        }

        public double MinorStart(Rect rect)
        {
            return IsVerical ? rect.Left : rect.Top;
        }

        public void SetMajorSize(ref Rect rect, double value)
        {
            if (IsVerical)
            {
                rect.Height = value;
            }
            else
            {
                rect.Width = value;
            }
        }

        public void SetMajorStart(ref Rect rect, double value)
        {
            if (IsVerical)
            {
                rect.Y = value;
            }
            else
            {
                rect.X = value;
            }
        }

        public void SetMinorStart(ref Rect rect, double value)
        {
            if (IsVerical)
            {
                rect.X = value;
            }
            else
            {
                rect.Y = value;
            }
        }

        public Rect MinorMajorRect(double minor, double major, double minorSize, double majorSize)
        {
            return
                IsVerical ?
                new Rect(minor, major, minorSize, majorSize) :
                new Rect(major, minor, majorSize, minorSize);
        }

        public Point MinorMajorPoint(double minor, double major)
        {
            return
                IsVerical ?
                new Point(minor, major) :
                new Point(major, minor);
        }

        public Size MinorMajorSize(double minor, double major)
        {
            return
                IsVerical ?
                new Size(minor, major) :
                new Size(major, minor);
        }
    }
}

File: dev/Repeater/OrientationBasedMeasures.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

#include <pch.h>
#include "OrientationBasedMeasures.h"

float& OrientationBasedMeasures::Major(const winrt::Size &size)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Size&)size).Height : ((winrt::Size&)size).Height;
}

float& OrientationBasedMeasures::Minor(const winrt::Size& size)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Size&)size).Width : ((winrt::Size&)size).Height;
}

float& OrientationBasedMeasures::MajorSize(const winrt::Rect& rect)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Rect&)rect).Height : ((winrt::Rect&)rect).Width;
}

float& OrientationBasedMeasures::MinorSize(const winrt::Rect& rect)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Rect&)rect).Width : ((winrt::Rect&)rect).Height;
}

float& OrientationBasedMeasures::MajorStart(const winrt::Rect& rect)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Rect&)rect).Y : ((winrt::Rect&)rect).X;
}

float OrientationBasedMeasures::MajorEnd(const winrt::Rect& rect) const
{
    return m_orientation == ScrollOrientation::Vertical ?
        rect.Y + rect.Height : rect.X + rect.Width;
}

float& OrientationBasedMeasures::MinorStart(const winrt::Rect& rect)
{
    return m_orientation == ScrollOrientation::Vertical ? ((winrt::Rect&)rect).X : ((winrt::Rect&)rect).Y;
}

float OrientationBasedMeasures::MinorEnd(const winrt::Rect& rect) const
{
    return m_orientation == ScrollOrientation::Vertical ?
        rect.X + rect.Width : rect.Y + rect.Height;
}

winrt::Rect OrientationBasedMeasures::MinorMajorRect(float minor, float major, float minorSize, float majorSize)
{
    return m_orientation == ScrollOrientation::Vertical ?
        winrt::Rect{ minor, major, minorSize, majorSize } :
        winrt::Rect{ major, minor, majorSize, minorSize };
}

winrt::Point OrientationBasedMeasures::MinorMajorPoint(float minor, float major)
{
    return m_orientation == ScrollOrientation::Vertical ?
        winrt::Point(minor, major) :
        winrt::Point(major, minor);
}

winrt::Size OrientationBasedMeasures::MinorMajorSize(float minor, float major)
{
    return m_orientation == ScrollOrientation::Vertical ?
        winrt::Size(minor, major) :
        winrt::Size(major, minor);
}
